### ISSUE TITLE
How to handle amalgamated libraries when building library statically 

### ISSUE THREAD

**Issue by gtalis** [amzn/hawktracer#75]
I have used your library in one of my projects and now I am trying to integrate it into the buildroot (https://buildroot.org) build system.
Everything works fine apart from the amalgamated libraries: they are built as shared libraries even though the main library is compiled statically.
When the project is cross-compiled using a full static toolchain with no support for shared libraries, the amalgamated libraries fail to build.
A couple of questions:
- what is the usage of these libraries, compared to the main one? I failed to understand the purpose of having them.
- what is the best way to handle them when trying to build the project in a full static way? Ignore them or building them as static libraries as well?

thanks for your help and this great project.

**Comment by andreasbuhr** [comment]
Thanks for your effort to integrate HawkTracer into buildroot.
The purpose of the amalgamated libraries is mainly the generation of a hawktracer.h and a hawktracer.{c|cpp} file wherein all files of the HawkTracer libraries are blended together. Have a look at the generated hawktracer.cpp in your build directory to understand how it works. Using these files, it is easy to add HawkTracer to a project without building HawkTracer as a library, but instead copying these files into the project.
The compilation of the amalgamated hawktracer.{h|c|cpp} files into a shared library is mainly to test that the amalgamation worked. So regarding your second question: You can safely ignore the amalgamated libraries for your use case.

Does this make sense?

**Comment by gtalis** [comment]
Hello Andreas,
yes, it does make sense. Thanks for the feedback and clear explanation.
Could we think of making the amalgamation step conditional using a cmake option?
If you are ok with this idea, I can send a pull request with the amalgamation option, set on ON by default.
thanks!
Gilles.

**Comment by loganek** [comment]
Hi @gtalis 
I think having the step optional make sense. 
Just out of curiosity, what was the error you experienced? Even though we make the step optional, I think the build shouldn't be failing, so it'd be good to investigate that failure. 

Many thanks for your support :)

**Comment by gtalis** [comment]
Hello,
I have attached the build failure. One of the compliance tests required to push a new package into buildroot is to build that package using different toolchains. Some of these toolchains only support static libraries, not shared libraries. **Obviously, using such a toolchain is a rare case**,but it often allows detecting compilation errors around static linking of libraries.
[logfile.txt](https://github.com/amzn/hawktracer/files/4598394/logfile.txt)


As you can see in the error log, libhawktracer builds fine, but the amalgamated libraries fail to build.

**Comment by loganek** [comment]
@gtalis thanks a lot for your change. After looking at the logs I understood what the problem is.
I think there's nothing wrong with having amalgamated build optional, but I think we could actually bring more value if we actually run it for every build. However, to address the issue you experienced, we could consider changing the amalgamation build to be dependent on `BUILD_STATIC_LIB`.
We already store the library type in `HAWKTRACER_LIB_TYPE` variable ([here](https://github.com/amzn/hawktracer/blob/master/CMakeLists.txt#L82)), so in the [test_amalgamation.cmake](https://github.com/amzn/hawktracer/blob/master/cmake/test_amalgamation.cmake) file we could replace `SHARED` with `${HAWKTRACER_LIB_TYPE}` (exactly as we do in [lib/CMakeLists.txt](https://github.com/amzn/hawktracer/blob/master/lib/CMakeLists.txt#L117)).
We'd probably also have to reorder [the line](https://github.com/amzn/hawktracer/blob/master/CMakeLists.txt#L68) with definition of `HAWKTRACER_LIB_TYPE` so the variable is defined before we include amalgamation cmake file.

Do you think that would work for you? Let me know if you're happy with that and you're ok with updating your pull requests. As I said, your PR looks good, I just think we could have even more value with always building amalgamated library so we can validate it (please note it's already optional, if python is not available in the environment).

**Comment by gtalis** [comment]
@loganek, thanks for your reply. Actually, your suggestion was the first one I had thought of, but I did not know if that would meet your needs or break something else in the package. Let me go ahead and make that change then re-submit my pull request. Thanks for the review!

**Comment by loganek** [comment]
Thank you @gtalis Also, could you please tell us how can we track the work you do for integrating hawktracer to buildroot? We'd like to make the announcement for HawkTracer's community once the work is done:) Thanks

**Comment by gtalis** [comment]
@loganek, I have submitted a pull request following your suggestion of building amalgamated libs with same type as hawktracer lib. Whilst testing the patch, I bumped into a new error further down the road, reported by the buildroot test suite. I am currently checking it and will open a new issue if required. Thanks!

**Comment by loganek** [comment]
@gtalis thank you. The PR looked great so I merged it. I'm closing this issue as it seems to be resolved. Feel free to open a new one if you find another problem; also please keep us updated about the progress of integrating HawkTracer to build root, really looking forward to see it!
